### DISCUSSION TITLE
Require an explicit mode for weather integration and bucket into three modes

### DISCUSSION THREAD

**Discussion by MatthewFlamm** [home-assistant/architecture#781] [Entity Models] [answered]
## Context

Currently, the weather integration supports hourly and daily weather forecasts.  Currently, the differentiator is whether a low temperature in `templow` is supplied by the backend and or how often the forecast is supplied.  This is cumbersome for the frontend to determine what kind of display is needed based on the availability and structure of the data.  It can also leave to issues such as this [one](https://github.com/home-assistant/frontend/issues/5792) where the structure of the backend data wasn't expected despite being a once daily forecast.  The fix required checking multiple forecasts to infer the intent.

There are currently three broad forecast types.  One is a once daily with a low temperature, the other is 'hourly' which is the type for any other forecast besides daily.  There is a middle class of forecasts that are not served well by either these two types: twice daily forecasts, ie a separate forecast for day and a separate forecast for night.  This causes issues like [this](https://github.com/home-assistant/frontend/issues/5829) where the displayed forecasts are hard to understand when it occurs.  It is more natural to differentiate by the date and whether the forecast is for the day or night, e.g. Sun and Sun Night.  But the current system treats it as hourly.

## Proposal

Require weather entities to supply a `mode` attribute that can have one of three values: `hourly`, `daily`, `daynight`.  `daily` forecasts can optionally supply a `templow` value for forecasts.  `daynight` forecasts will require a `is_daytime` boolean to differentiate.  `hourly` forecasts will work the same as the current state.

## Consequences

It is easier to display data from weather integrations and avoid issues when subtle differences in data structure and type arise.  It requires all weather integrations to add one more attribute and the frontend card to use the new attribute to display the data.

**Comment by zsarnett** [comment]
Working with the Weather Card recently this was a very big problem that can be easily solved by the backend supplying which forecast is being sent. Specifically for DarkSky you actually have to set it in your config. If that was sent as an attribute on the entity it would make creating an agnostic weather card much easier.

**Comment by zsarnett** [comment]
Minimal changes I believe should be supplying the mode. `hourly`, `daily`, and `daynight`.

**Comment by zsarnett** [comment]
https://github.com/home-assistant/frontend/pull/6065

Another issue this would resolve if the Backend Supplied the mode. Instead of fixing each one a different way using guessing in the frontend we can fix them all and make the frontend code less complicated with this approach.

**Comment by MatthewFlamm** [comment]
> Minimal changes I believe should be supplying the mode. `hourly`, `daily`, and `daynight`.


I think this suggestion just removes the required `is_daytime` boolean for `daynight` entities. 
How will the frontend know which is which?  It isn't clear to me.  The timestamps are confusing in this case as Monday night forecast predominately occurs in Tuesday.  This last part is why the current timestamp for NWS daynight entity is the start time of the forecast.

**Comment by rajlaud** [comment]
I also recommend having an `is_daytime` boolean. In the NWS forecasts, for example, the first forecast is short term, rather than at a fixed time, followed by 6:00 a.m. and 6:00 p.m. forecasts. To correctly distinguish day time and night time forecasts using timestamps, 6:00 p.m. precisely needs to be the dividing line. Other forecasts might be different, and the frontend would have to know them all.

**Comment by MatthewFlamm** [comment]
nws is currently the only integration offering a day and night forecast, but environment_canada is discarding it's night forecast to force it to be a daily forecast [here](https://github.com/home-assistant/core/blob/da6a99b333dfe939f8d16f017b3ffa85fd30f52b/homeassistant/components/environment_canada/weather.py#L176-L204), so it could probably support this mode as well.  Pinging @michaeldavie to confirm whether this is true.

**Comment by michaeldavie** [comment]
Yes, it would actually be simpler for me to handle the overnight forecasts as equivalent to the daytime ones. My backend library already does this, and I have to do some contortions to shoehorn the data into the form Home Assistant expects.

This would also provide more data, as I'm only keeping the overnight temperature at the moment and discarding the other forecast conditions.

**Comment by MatthewFlamm** [comment]
https://github.com/home-assistant/core/pull/38725 adds another integration with day+night forecasts.

**Comment by zsarnett** [comment]
I would like us to decide a path for this as this will help shape the front end code

**Comment by Makin-Things** [comment]
So 2 years on and it appears no architectural decision has been made on this? But the NWS integration has just gone ahead and implemented it as if it was approved. The developer documentation hasn't been updated so yes/no/maybe its approved!?
Asking because I am writing a new weather card and would like to support as many users as possible, but I need to understand what I am trying to support.

**Comment by frenck** [comment]
>  But the NWS integration has just gone ahead and implemented it as if it was approved.

It is implemented as a local solution for an integration, not as a global architectural change. This proposal has not been approved or implemented at this point.

**Comment by Makin-Things** [comment]
> This proposal has not been approved or implemented at this point.

But the NWS integration which is in core has implemented it and it has driven a change though to the forecast card.

**Comment by frenck** [comment]
@Makin-Things There is no architectural global change made as proposed in this issue.

**Comment by Makin-Things** [comment]
Also I am not trying to be confrontational, but it took a long time to find out why the NWS integration was doing things differently.
The first question though is this the best place to be discussing this?

**Comment by frenck** [comment]
> Also I am not trying to be confrontational, but it took a long time to find out why the NWS integration was doing things differently.

That is possible, doesn't mean it standardized.

> The first question though is this the best place to be discussing this?

Your original question was about this architectural proposal, which has been answered, as its not applied.

If you have other questions (not directly related to discussing this proposal), please use our community forums or Discord chat.

**Comment by emontnemery** [comment, marked as answer]
## Suggestion:

Instead of adding a mode on the entity level, add new attributes `forecast_hourly` + `forecast_daily` + `forecast_twice_daily`, which the entity can fill depending on what's received from the underlying service. For `forecast_twice_daily` entries it is mandatory to set the boolean attribute `is_daytime`.
This would then allow displaying the data like this (for example):
![image](https://user-images.githubusercontent.com/14281572/177850822-b94405ee-2679-43b8-be7c-c7bf2880dea2.png)

**Reply by MatthewFlamm** [reply]
This will work for integrations that have daily and hourly forecasts.  For integrations with separate day+night forecasts, we will have to either throw out half the forecast, except for the temperature, or provide two hourly entities, one of which only has two entries per day.

**Reply by Makin-Things** [reply]
I guess this could also work. @MatthewFlamm there could be a 3rd attribute ``forecast_twice_daily``. The one drawback I see with this approach is that if an integration provide both hourly and daily/twice_daily attributes you will get all the data in the attributes even if you are not going to use it. I think a lot of the weather integrations create the hourly entity, but leave it disabled by default.

**Reply by emontnemery** [reply]
If there are many integrations which could make use of it, we could include a `forecast_twice_daily` or similar. How many integrations do support that?

**Reply by Makin-Things** [reply]
The only ones I am aware of are NWS (USA) and Env Canada (Canada). I am guessing that represents a fair percentage of the HA user base.

**Reply by MatthewFlamm** [reply]
NWS, Env Canada, and MetOffice all provide one day and one night forecast.  I haven't done a review of the new integrations since I opened this issue 2 years ago, so possibly more.

**Reply by gjohansson-ST** [reply]
I would suggest that we go with @emontnemery 's suggestion here to create the different attributes of forecast and given the input on a bi-weekly I would think all three seems to be a good fit.

**Reply by emontnemery** [reply]
It seems this proposal is the favored one, I'll mark it as the answer

**Reply by Makin-Things** [reply]
Can we please clarify what the accepted solution is, considering there are multiple opinions that have put forward.

**Reply by emontnemery** [reply]
The accepted solution is to add new attributes `forecast_hourly` + `forecast_daily` + `forecast_twice_daily`. For `forecast_twice_daily` entries it is mandatory to set the boolean attribute `is_daytime`.

**Reply by emontnemery** [reply]
There was some late feedback in the core PR. The accepted solution has been slightly changed:
- Forecasts are not exposed as entity attributes, they are instead subscribed to via websocket
- Weather entities need to implement async methods `async_forecast_daily`, `async_forecast_twice_daily`, `async_forecast_hourly`
- Forecast support is signaled by setting supported features from feature enum `WeatherEntityFeature`

The rationale for the changes is to avoid weather platforms doing unnecessary API calls to fetch forecasts which may not be used.

**Comment by Makin-Things** [comment]
From the frontends point of view I would like to see an attribute added at the top level of the entity called something like ``forecast_mode`` with possible values of ``daily``, ``hourly`` or ``twice_daily``. Not really fussed about the actual values, but something that clearly states what is in the included in the ``forecast`` blob. At present there isn't even a simple way to differentiate between ``daily`` and ``hourly``, you have to compare timestamps in the forecast entries to figure it out.

@MatthewFlamm for the NWS with its twice daily forecasts did you consider keeping both of each days forecasts under the same forecast[x] but with an object/array then containing the 2 (morning/evening) forecasts.

**Reply by MatthewFlamm** [reply]
> @MatthewFlamm for the NWS with its twice daily forecasts did you consider keeping both of each days forecasts under the same forecast[x] but with an object/array then containing the 2 (morning/evening) forecasts.

IMO this adds too much hierarchical complexity.  But, if that is preferred then I'm okay with it.

**Reply by Makin-Things** [reply]
Yes it is a little more complex, but it would mean that the data would be in a deterministic location. For example tomorrows minimum temperature would always be at ``forecasst[1].morning.temperature``.
At the moment you have to search through all the data to find what you are looking for as ``forecast[0]` can be either morning or afternoon.

**Reply by michaeldavie** [reply]
Environment Canada also has hourly and half-day forecasts, with the added bonus of only having the evening forecast for the current day once it’s the afternoon. 

I think it makes sense to have separate forecast objects that refer to a single time period, and noting somehow what that time period is.

**Reply by MatthewFlamm** [reply]
Edit: this was in reference to

> For example tomorrows minimum temperature would always be at forecasst[1].morning.temperature.

This is a good point, I'm in favor of this suggestion as an option too.

**Reply by emontnemery** [reply]
This basically describes the original proposal by @MatthewFlamm. It seems less flexible than allowing the entity to expose both hourly and daily (and possibly bi-daily) forecasts in the same entity, and will not encourage integrations to not expose multiple entities or remove the need for users to configure entities for daily or hourly forecast, and it will not allow an advanced weather card as in the screenshot above.

Why do you think this way is better for the frontend @MatthewFlamm ?

**Reply by MatthewFlamm** [reply]
> Why do you think this way is better for the frontend @MatthewFlamm ?

I still think my proposal is better since it provides a uniform structure for each forecast type.  It still makes it difficult to find a specific forecast, but this is a downside for any forecast type, not just the day+night forecasts.

The proposal of squeezing both forecasts into one object would be easier to find a specific forecast.  This is certainly useful, so I'm okay with it too if that is what others prefer.  My opinion is that this proposal would be better if there is some logical structure that could be used for each forecast type, not just the day+night forecast.  If that is possible, then I might switch my vote.

**Reply by Makin-Things** [reply]
This is my 2c worth from someone that is trying to develop a card that supports all weather integrations
I have some testers that use the NWS integration and have to do all sorts of template gymnastics to massage the twice daily forecasts to make things work for them.
I have been putting together a new weather card that is just a little bit customisable ([https://github.com/Makin-Things/platinum-weather-card](url))
The twice daily forecasts currently don't work well, but I would love to be able to improve that (deterministic data would make it much easier).
I guess I am asking to share the pain of inconsistent met data a little bit. Frontend cards need to know what they are dealing with (or the user needs to create a bunch of template entities to work around the deficiencies).

**Comment by Makin-Things** [comment]
This is a little OT, but I would also like to see some additional optional attributes added to the forecasts (this is more just a documentation thing so authors know what to name things).
For example the NWS integration has added ``detailed_description``. Because this is not in the documentation the next person that decides to add a text forecast to an integration will call it something else, such as ``text_forecast``, ``extended_text`` etc.
I would also like to see something defined for a short version of the forecast. The BoM (Australia) and most other met offices tend to have a headline/short forecast for the day consisting of up to about 4 words. As an example this is what my weather looks like right now.
![image](https://user-images.githubusercontent.com/91187456/177889577-9c4b14cc-c9e8-49b7-9a36-5664aac99d45.png)
I'd also like to see ``condition_raw`` added that does no translation of the real data. Using condition I get told it is going to snow when there is just going to be a frost.

**Reply by michaeldavie** [reply]
Maybe? For Environment Canada I use sensors for most of the details, of which there are quite a few.

**Reply by Makin-Things** [reply]
The BoM integration also creates lots (and I mean lots ie. 150+). That only includes the current observations and the next 7 daily forecasts. It doesn't have any for the hourly forecasts which is where the condition_raw would be useful (the other 2 are only relevant for the daily forecast here).
I also forgot that having an apparent/feels like temperature at the top level would also be useful. The BoM publishes this value, but I believe there is a formula that can be used to calculate it.

**Reply by frenck** [reply]
> I'd also like to see condition_raw added that does no translation of the real data. Using condition I get told it is going to snow when there is just going to be a frost.

This could already be passed as an extra attribute. That said, we generally avoid these type of things, as they end up in the state machine. The state machine is active in memory, has changes recorded, and is thus not really good/suitable for storing blobs of text.

**Comment by Makin-Things** [comment]
I also think there is an error in the documentation here:
https://developers.home-assistant.io/docs/core/entity/weather/
All of the weather integrations I've looked at have the current condition as the state of the entity, the documentation says it's a required attribute at the top level, which I have never seen.

**Reply by emontnemery** [reply]
You're confusing the state object and its state and attributes with the entity property methods.
The weather entity generates the state of the state object like so:
```py
    @property
    @final
    def state(self) -> str | None:
        """Return the current state."""
        return self.condition
```

I'm hiding this thread as off-topic.

**Comment by Makin-Things** [comment]
Also OT, but I raised this issue as well earlier today.
https://github.com/home-assistant/frontend/issues/13145
Since the integration may be creating entities in a different time zone there is a need to have an attribute that has the timezone stored so that the displayed time can be shown in the entities actual timezone.

**Reply by emontnemery** [reply]
Please open a new discussion suggesting such an attribute, let's keep this discussion about the forecast modes 👍

**Reply by Makin-Things** [reply]
No problem. Is there a good way to do that and get the right audience? I suspect this topic has a good bunch of people that are interested/knowledgeable on the topic.

**Reply by frenck** [reply]
> I suspect this topic has a good bunch of people that are interested/knowledgeable on the topic.

It also pollutes the discussion. Please, don't do that. Thanks 👍

**Comment by Makin-Things** [comment]
@frenck as I said, not a problem, but I also asked for guidance on how to start a new discussion and get the right people included. I am new here but have a lot to offer, so some help rather than don't do that would be appreciated.

**Reply by frenck** [reply]
>  I am new here but have a lot to offer, so some help rather than don't do that would be appreciated.

Welcome 👋 ! And don't worry about it.

You have raised an issue in the frontend issue tracker 3 hours ago. Just have some patience. No need to paste that link around to get attention to it.

**Comment by Makin-Things** [comment]
Ok. no problem. Happy for you to hide this as well. I'd like to start a more general discussion about the weather stuff that includes the various things I raised and noted as OT. What is the best way to do that?

**Comment by Makin-Things** [comment]
Completely back on topic. I would really like to see the forecast data being able to be found in a deterministic way. I don't really care what the method employed to do that is, but it really needs to be deterministic.
